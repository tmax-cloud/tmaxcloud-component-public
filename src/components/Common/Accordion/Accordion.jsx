import PropTypes from "prop-types";
import { Accordion_Box, Accordion_List, Accordion_Row } from "./";

const Accordion = (props) => {
  if (props.state === "Box") return <Accordion_Box {...props} />;
  else if (props.state === "List") return <Accordion_List {...props} />;
  else return <Accordion_Row {...props} />;
};
Accordion.propTypes = {
  /** 아코디언 타입 */
  state: PropTypes.oneOf(["Box", "List", "Row"]),
  /** 아코디언 카테고리(state가 List인 경우 활성 사용 가능) */
  category: PropTypes.string,
  /** 아코디언의 제목  */
  title: PropTypes.string,
  /** 아코디언의 내용 */
  content: PropTypes.string,
  /** 이용불가능 여부 */
  disabled: PropTypes.bool,
};
Accordion.defaultProps = {
  state: "Box",
  category: "카테고리",
  title: "타이틀",
  content: "내용물",
  disabled: false,
};

export default Accordion;
