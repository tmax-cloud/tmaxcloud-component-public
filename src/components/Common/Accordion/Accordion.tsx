import { Accordion_Box, Accordion_List, Accordion_Row } from ".";
type AccordionPropsType = {
  state: "Box" | "List" | "Row";
  /** 카테고리(state가 List인 경우 활성 사용 가능) */
  category: string;
  /** 제목  */
  title: string;
  /** 내용 */
  content: string;
  /** 이용 불가 여부 */
  disabled: boolean;
};
/** state에 따른 아코디언 분류 컴포넌트 */
const Accordion = (props: AccordionPropsType) => {
  if (props.state === "Box") return <Accordion_Box {...props} />;
  else if (props.state === "List") return <Accordion_List {...props} />;
  else return <Accordion_Row {...props} />;
};

Accordion.defaultProps = {
  state: "Box",
  category: "카테고리",
  title: "타이틀",
  content: "내용물",
  disabled: false,
};

export default Accordion;
