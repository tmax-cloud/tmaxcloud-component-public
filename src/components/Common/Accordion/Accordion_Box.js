import { useState } from "react";
import PropTypes from "prop-types";
import styled, { css } from "styled-components";

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  border-bottom: 1px solid ${({ theme }) => theme.color.gray._300};
`;
const TitleArea = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  cursor: pointer;
  width: 100%;
  height: 4.8rem;
  padding-left: 2rem;
  padding-right: 2rem;
  ${({ theme }) => theme.font.body2_400};
  background-color: ${({ theme, isClick }) =>
    isClick ? theme.color.gray._50 : theme.color.white._100};

  :hover {
    background-color: ${({ theme }) => theme.color.gray._50};
  }
  :active {
    background-color: ${({ theme }) => theme.color.gray._100};
  }
  ${({ disabled }) =>
    disabled &&
    css`
      color: ${({ theme }) => theme.color.gray._900};
      opacity: 0.4;
      &,
      :hover,
      :active {
        background-color: ${({ theme }) => theme.color.gray._400};
      }
    `};
`;
const Icon = styled.div`
  width: 2.4rem;
  margin-left: 2rem;
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;
const Content = styled.div`
  width: 100%;
  padding: 0 2rem 1rem 2rem;
  background-color: ${({ theme }) => theme.color.gray._50};
  span {
    ${({ theme }) => theme.font.body3_400};
    color: ${({ theme }) => theme.color.gray._700};
  }
`;
const Accordion_Box = ({ title, content, disabled }) => {
  const [click, setClick] = useState(false);
  const onClick = () => {
    if (!disabled) {
      setClick(!click);
    }
  };
  return (
    <Wrapper>
      <TitleArea onClick={onClick} isClick={click} disabled={disabled}>
        <span>{title}</span>
        <Icon>
          {click ? (
            <img
              src="asset/images/Icon/arrow/Direction=up, Size=Medium.svg"
              alt="close"
            />
          ) : (
            <img
              src="asset/images/Icon/arrow/Direction=down, Size=Medium.svg"
              alt="open"
            />
          )}
        </Icon>
      </TitleArea>
      {click && (
        <Content>
          <span>{content}</span>
        </Content>
      )}
    </Wrapper>
  );
};
Accordion_Box.propTypes = {
  title: PropTypes.string,
  content: PropTypes.string,
  disabled: PropTypes.bool,
};
Accordion_Box.defaultProps = {
  title: "타이틀",
  content: "내용물",
  disabled: false,
};

export default Accordion_Box;
