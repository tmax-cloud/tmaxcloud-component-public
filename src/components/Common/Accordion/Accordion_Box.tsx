import { useState } from "react";
import styled, { css } from "styled-components";
/** 박스형 아코디언 props 타입 */
type Accordion_BoxPropsType = {
  /** 타이틀 */
  title: string;
  /** 내용 */
  content: string;
  /** 이용 불가 여부 */
  disabled: boolean;
};

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  border-bottom: 1px solid ${({ theme }) => theme.color.gray._300};
`;
const TitleArea = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  cursor: pointer;
  width: 100%;
  height: 4.8rem;
  padding-left: 2rem;
  padding-right: 2rem;
  ${({ theme }) => theme.font.body2_400};
  background-color: ${({ theme, isClick }) =>
    isClick ? theme.color.gray._50 : theme.color.white._100};

  :hover {
    background-color: ${({ theme }) => theme.color.gray._50};
  }
  :active {
    background-color: ${({ theme }) => theme.color.gray._100};
  }
  ${({ disabled }) =>
    disabled &&
    css`
      color: ${({ theme }) => theme.color.gray._900};
      opacity: 0.4;
      &,
      :hover,
      :active {
        background-color: ${({ theme }) => theme.color.gray._400};
      }
    `};
`;
const Icon = styled.div`
  width: 2.4rem;
  margin-left: 2rem;
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;
const Content = styled.div`
  width: 100%;
  padding: 0 2rem 1rem 2rem;
  background-color: ${({ theme }) => theme.color.gray._50};
  span {
    ${({ theme }) => theme.font.body3_400};
    color: ${({ theme }) => theme.color.gray._700};
  }
`;
/** 박스형 아코디언 */
const Accordion_Box = ({
  title,
  content,
  disabled,
}: Accordion_BoxPropsType) => {
  const [click, setClick] = useState(false);
  /** 클릭이벤트 */
  const onClick = (): void => {
    if (!disabled) {
      setClick(!click);
    }
  };
  return (
    <Wrapper>
      <TitleArea onClick={onClick} isClick={click} disabled={disabled}>
        <span>{title}</span>
        <Icon>
          {click ? (
            <img src="asset/images/Icon/dummy_icon.svg" alt="close" />
          ) : (
            <img src="asset/images/Icon/dummy_icon.svg" alt="open" />
          )}
        </Icon>
      </TitleArea>
      {click && (
        <Content>
          <span>{content}</span>
        </Content>
      )}
    </Wrapper>
  );
};

Accordion_Box.defaultProps = {
  title: "타이틀",
  content: "내용물",
  disabled: false,
};

export default Accordion_Box;
