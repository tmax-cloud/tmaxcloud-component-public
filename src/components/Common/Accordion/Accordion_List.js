import { useState } from "react";
import PropTypes from "prop-types";
import styled, { css } from "styled-components";

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  border-bottom: 1px solid ${({ theme }) => theme.color.gray._300};
`;
const TitleArea = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  cursor: pointer;
  width: 100%;
  height: 9.6rem;
  padding-left: 2rem;
  padding-right: 2rem;
  padding-bottom: 1rem;
  background-color: ${({ theme, isClick }) =>
    isClick ? theme.color.gray._50 : theme.color.white._100};
  :hover {
    background-color: ${({ theme }) => theme.color.gray._50};
  }
  :active {
    background-color: ${({ theme }) => theme.color.gray._100};
  }
  ${({ disabled }) =>
    disabled &&
    css`
      color: ${({ theme }) => theme.color.gray._900};
      opacity: 0.4;
      &,
      :hover,
      :active {
        background-color: ${({ theme }) => theme.color.gray._400};
      }
    `};
`;
const Category = styled.div`
  /* 500이 없음 */
  color: ${({ theme }) => theme.color.gray._600};
  ${({ theme }) => theme.font.detail2_400}
  margin-bottom:5px;
`;
const Title = styled.div`
  ${({ isClick, theme }) =>
    isClick ? theme.font.body2_500 : theme.font.body2_400}
`;
const Icon = styled.div`
  width: 2.4rem;
  margin-left: 2rem;
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;
const Content = styled.div`
  width: 100%;
  padding: 0 2rem 1rem 2rem;
  background-color: ${({ theme }) => theme.color.gray._50};
  span {
    ${({ theme }) => theme.font.body2_400};
    color: ${({ theme }) => theme.color.gray._700};
  }
`;
const Accordion_List = ({ category, title, content, disabled }) => {
  const [click, setClick] = useState(false);
  const onClick = () => {
    !disabled && setClick(!click);
  };
  return (
    <Wrapper>
      <TitleArea onClick={onClick} isClick={click} disabled={disabled}>
        <div>
          <Category>{category}</Category>
          <Title isClick={click}>{title}</Title>
        </div>
        <Icon>
          {click ? (
            <img
              src="asset/images/Icon/arrow/Direction=up, Size=Medium.svg"
              alt="close"
            />
          ) : (
            <img
              src="asset/images/Icon/arrow/Direction=down, Size=Medium.svg"
              alt="open"
            />
          )}
        </Icon>
      </TitleArea>
      {click && (
        <Content>
          <span>{content}</span>
        </Content>
      )}
    </Wrapper>
  );
};
Accordion_List.propTypes = {
  category: PropTypes.string,
  title: PropTypes.string,
  content: PropTypes.string,
  disabled: PropTypes.bool,
};
Accordion_List.defaultProps = {
  category: "카테고리",
  title: "타이틀",
  content: "내용물",
  disabled: false,
};

export default Accordion_List;
