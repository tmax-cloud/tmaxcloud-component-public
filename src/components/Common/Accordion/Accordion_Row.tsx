import { useState } from "react";
import styled, { css } from "styled-components";
/** 줄형 아코디언 props 타입 */
type Accordion_RowPropsType = {
  /** 타이틀 */
  title: string;
  /** 내용 */
  content: string;
  /** 이용 불가 여부 */
  disabled: boolean;
};
const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  position: relative;
  ${({ isClick, theme }) =>
    isClick &&
    css`
      &::before {
        width: 0.2rem;
        height: 100%;
        position: absolute;
        top: 0;
        left: 0;
        content: "";
        background-color: ${theme.color.gray._900};
      }
    `};
`;
const TitleArea = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  height: 48px;
  padding-left: 2rem;
  padding-right: 2rem;
  border-radius: 0.4rem;
  cursor: pointer;
  ${({ theme }) => theme.font.body2_400};
  ${({ isClick, theme }) =>
    isClick ? theme.font.body2_500 : theme.font.body2_400}
  :hover {
    background-color: ${({ theme }) => theme.color.gray._50};
  }
  :active {
    background-color: ${({ theme }) => theme.color.gray._100};
  }
  ${({ disabled }) =>
    disabled &&
    css`
      color: ${({ theme }) => theme.color.gray._900};
      opacity: 0.4;
      &,
      :hover,
      :active {
        background-color: ${({ theme }) => theme.color.gray._400};
      }
    `};
`;
const Icon = styled.div`
  width: 24px;
  margin-left: 2rem;
  display: inline-flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
`;
const Content = styled.div`
  width: 100%;
  padding: 0 2rem 1rem 2rem;
  background-color: ${({ theme }) => theme.color.white._100};

  span {
    ${({ theme }) => theme.font.body3_400};
    color: ${({ theme }) => theme.color.gray._700};
  }
`;
/**줄형 아코디언 */
const Accordion_Row = ({
  title,
  content,
  disabled,
}: Accordion_RowPropsType) => {
  const [click, setClick] = useState(false);
  const onClick = () => {
    !disabled && setClick(!click);
  };
  return (
    <Wrapper isClick={click}>
      <TitleArea onClick={onClick} isClick={click} disabled={disabled}>
        <span>{title}</span>
        <Icon>
          {click ? (
            <img src="asset/images/Icon/dummy_icon.svg" alt="close" />
          ) : (
            <img src="asset/images/Icon/dummy_icon.svg" alt="open" />
          )}
        </Icon>
      </TitleArea>
      {click && (
        <Content>
          <span>{content}</span>
        </Content>
      )}
    </Wrapper>
  );
};

Accordion_Row.defaultProps = {
  title: "타이틀",
  content: "내용물",
  disabled: false,
};

export default Accordion_Row;
