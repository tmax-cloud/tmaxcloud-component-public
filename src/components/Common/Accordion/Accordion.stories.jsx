import React from "react";
import Accordion from "./Accordion";

export default {
  title: "Common/Accordion",
  component: Accordion,
  argTypes: {},
  decorators: [
    (Story) => (
      <div>
        <Story />
      </div>
    ),
  ],
  parameters: {
    componentSubtitle:
      "제한된 공간에 정보를 표시하기 위해 접을 수 있는 내용 패널",
    docs: {
      description: {
        component: "Accordion",
      },
    },
  },
};

const Template = (args) => <Accordion {...args} />;

export const AccordionTemplate = Template.bind({});

AccordionTemplate.args = {
  state: "Box",
  title: "타이틀",
  content: "내용물",
  disabled: false,
};
