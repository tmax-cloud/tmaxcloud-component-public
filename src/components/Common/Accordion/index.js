import Accordion_Box from "./Accordion_Box";
import Accordion_List from "./Accordion_List";
import Accordion_Row from "./Accordion_Row";

export { Accordion_Box, Accordion_Row, Accordion_List };
