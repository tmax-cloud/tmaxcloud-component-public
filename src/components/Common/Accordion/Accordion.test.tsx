import Accordion from "./Accordion";
import { render, screen } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import { defaultTheme } from "styles/theme";
import userEvent from "@testing-library/user-event";
describe("Accordion 공통 테스트", () => {
  it("제목 클릭에 따른 하단 내용 노출 여부 (미클릭, 첫클릭, 두번 클릭)", () => {
    render(
      <ThemeProvider theme={defaultTheme}>
        <Accordion
          state="List"
          category="아코디언 테스트"
          title="테스트 제목"
          content="테스트 내용"
          disabled={false}
        />
      </ThemeProvider>,
    );
    /** 처음엔 하단 내용 노출 X */
    expect(() => screen.getByText("테스트 내용")).toThrow(Error);
    userEvent.click(screen.getByText("테스트 제목"));
    /** 클릭 후 하단 내용 노출 */
    expect(screen.getByText("테스트 내용")).toBeTruthy();
    userEvent.click(screen.getByText("테스트 제목"));
    /** 다시 클릭 후 하단 내용 노출 X */
    expect(() => screen.getByText("테스트 내용")).toThrow(Error);
  });

  it("disabled 활성화시 click 이벤트 정지 여부", () => {
    render(
      <ThemeProvider theme={defaultTheme}>
        <Accordion
          state="List"
          category="아코디언 테스트"
          title="테스트 제목"
          content="테스트 내용"
          disabled={true}
        />
      </ThemeProvider>,
    );
    userEvent.click(screen.getByText("테스트 제목"));
    expect(() => screen.getByText("테스트 내용")).toThrow(Error);
  });
});
describe("Accordion 타입별 카테고리 텍스트 노출 테스트", () => {
  it("Box Type 카테고리 노출 X", () => {
    render(
      <ThemeProvider theme={defaultTheme}>
        <Accordion
          state="Box"
          category="아코디언 테스트"
          title="테스트 제목"
          content="테스트 내용"
          disabled={true}
        />
      </ThemeProvider>,
    );
    consoasdfasdfle.log("error");
    expect(() => screen.getByText("아코디언 테스트")).toThrow(Error);
  });

  it("List Type 카테고리 노출 O", () => {
    render(
      <ThemeProvider theme={defaultTheme}>
        <Accordion
          state="List"
          category="아코디언 테스트"
          title="테스트 제목"
          content="테스트 내용"
          disabled={true}
        />
      </ThemeProvider>,
    );
    expect(() => screen.getByText("아코디언 테스트")).toBeTruthy();
  });

  it("Row Type 카테고리 노출 X", () => {
    render(
      <ThemeProvider theme={defaultTheme}>
        <Accordion
          state="Row"
          category="아코디언 테스트"
          title="테스트 제목"
          content="테스트 내용"
          disabled={true}
        />
      </ThemeProvider>,
    );
    expect(() => screen.getByText("아코디언 테스트")).toThrow(Error);
  });
});
