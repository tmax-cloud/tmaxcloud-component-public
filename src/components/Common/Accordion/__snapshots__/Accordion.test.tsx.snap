// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion 공통 테스트 matches snapshot 1`] = `
<div>
  <div
    class="sc-gKXOVf knFhxB"
  >
    <div
      class="sc-iBkjds gVxndz"
    >
      <div>
        <div
          class="sc-ftvSup cTnNJp"
        >
          아코디언 테스트
        </div>
        <div
          class="sc-papXJ QbWaf"
        >
          테스트 제목
        </div>
      </div>
      <div
        class="sc-jqUVSM bIizqJ"
      >
        <svg
          class="sc-crXcEl kmverq"
        >
          Medium.svg
        </svg>
      </div>
    </div>
  </div>
</div>
`;
